doc: Rewrite python onboarding tutorials in manual

bhipple opened the pull request:
Based on some feedback in #87094 and discussion with @fridh, this re-organizes
the onboarding tutorial in the Nixpkgs manual's python section, so that we start
with the simplest, most ad-hoc examples and work our way up. This progresses
from:

1. How to create an temporary python env at the cmdline, then
2. How to create a specific python env for a single script, then
3. How to create a specific python env for a project in a shell.nix, then
4. How to install a specific python env globally on the system or in a user profile.

Additionally, I've tried to standardize on some of the "best practice" ways of
doing things:

1. Instead of saying that this command style is "supported but strongly
   discouraged", I've just deleted it to avoid confusion.

   Bad:  nix-shell -p python38Packages.numpy python38Packages.toolz
   Good: nix-shell -p 'python38.withPackages(ps: with ps; [ numpy toolz ])'

2. In the portion where we show how to add stuff to the user's
   `XDG_CONFIG_HOME`, use overlays instead of `config.nix`. The former can do
   everything the latter can do, but is also much more generic and powerful,
   because it can compose with other files, compose with other envs, compose
   with overlays that do things like swap whether tensorflow and pytorch are
   built openblas/mkl/cuda stacks, and so on. The user is eventually going to
   see the overlay, so to avoid confusion let's standardize on it.

Let me know what you guys think!

bhipple committed:
doc: Rewrite python onboarding tutorials in manual

Based on some feedback in #87094 and discussion with @fridh, this re-organizes
the onboarding tutorial in the Nixpkgs manual's python section, so that we start
with the simplest, most ad-hoc examples and work our way up. This progresses
from:

1. How to create an temporary python env at the cmdline, then
2. How to create a specific python env for a single script, then
3. How to create a specific python env for a project in a shell.nix, then
4. How to install a specific python env globally on the system or in a user profile.

Additionally, I've tried to standardize on some of the "best practice" ways of
doing things:

1. Instead of saying that this command style is "supported but strongly not
   discouraged", I've just deleted it to avoid confusion.

   Bad:  nix-shell -p python38Packages.numpy python38Packages.toolz
   Good: nix-shell -p 'python38.withPackages(ps: with ps; [ numpy toolz ])'

2. In the portion where we show how to add stuff to the user's
   `XDG_CONFIG_HOME`, use overlays instead of `config.nix`. The former can do
   everything the latter can do, but is also much more generic and powerful,
   because it can compose with other files, compose with other envs, compose
   with overlays that do things like swap whether tensorflow and pytorch are
   built openblas/mkl/cuda stacks, and so on. The user is eventually going to
   see the overlay, so to avoid confusion let's standardize on it.

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
-  ];
-}
+Notice that this time it built a new python environment, which now includes
+`requests`. This is just creating wrapper scripts that setup `NIX_PYTHONPATH`,
```
I would avoid using details such as `NIX_PYTHONPATH`. It's not really important for the user at this point, and `NIX_PYTHONPATH` is an implementation detail.

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
+but re-using the actual modules themselves, so it's fast and efficient -- even
+if we're adding "big" modules like `pytorch` or `tensorflow`.
+
+To find the names of python attributes, startup `nix repl .` in a Nixpkgs
```
Cannot yet use the new `nix` experimental command, especially now because it goes behind a `experimental-features` option. Otherwise it would definitely be good to add.

bhipple replied:
That's really . . . unfortunate.  The `nix repl` is a much faster, nicer way of finding what the magic python attribute names are than the nixpkgs search on the web browser :(  

I'll remove it though.

FRidh commented on doc/languages-frameworks/python.section.md:
````diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
 ```
 
-opens a shell from which you can launch the interpreter
+Executing this script requires a python3 that has numpy. Using what we learned
````
`python3` and `numpy`

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
+system.
+
+By default it's pulling the import checkout of Nixpkgs itself from our nix
+channel, which is nice as it cache aligns with our other package builds, but we
```
its

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
+
+By default it's pulling the import checkout of Nixpkgs itself from our nix
+channel, which is nice as it cache aligns with our other package builds, but we
+can make it fully reproducible by pinning the `nixpkgs` import:
```
:+1:

FRidh commented on doc/languages-frameworks/python.section.md:
````diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
 ```
 
-Executing `nix-shell` gives you again a Nix shell from which you can run Python.
+And then at the cmdline, just typing `nix-shell` produces the same environment
````
command line

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -312,7 +433,7 @@ Our example, `toolz`, does not have any dependencies on other Python packages or
 arguments `buildInputs` and `propagatedBuildInputs` to specify dependencies. If
 something is exclusively a build-time dependency, then the dependency should be
-included as a `buildInput`, but if it is (also) a runtime dependency, then it
+included as a `buildInputs`, but if it is (also) a runtime dependency, then it
```
"included in" instead of "included as a"

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -423,10 +544,11 @@ Note also the line `doCheck = false;`, we explicitly disabled running the test-s
+(`python setup.py develop`); instead of installing the package this command
+creates a special link to the project code. That way, you can run updated code
+without having to reinstall after each and every change you make. Development
+mode is also available. Let's see how you can use it.
```
is also available for `setuptools`-based projects.

While we now have PEP 517/518, there is no generic approach (yet).

FRidh reviewed (changes requested):
This looks good to me, thank you for picking this up! Often there are remarks on documentation but it is not often someone actually does something about it.

I only have minor comments. One small change I would like to see is that "Python" is used throughout the text instead of "python".

danieldk commented on doc/languages-frameworks/python.section.md:
```diff
@@ -36,129 +36,202 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
+but re-using the actual modules themselves, so it's fast and efficient -- even
+if we're adding "big" modules like `pytorch` or `tensorflow`.
+
+To find the names of python attributes, startup `nix repl .` in a Nixpkgs
```
I guess `nix repl '<nixpkgs>'` would be simpler, since the user doesn't need a `nixpkgs` checkout and they'll see packages from their currently configured channel.

bhipple replied:
They need a channel, though.  Personally I've always found channels to be a confusing stumbling block for beginners; most developers can intuitively grasp a git repo clone, and git pull / checkout to change their "channel", and can reason about it: e.g., I'm getting "the exact python expressions from *this directory's* python versions".  

At any rate, it's irrelevant in this case, since it seems we're purging reference to nix 2.0 cmds in the manual for the time being.

bhipple commented:
(While we iterate on the reviews, I'll leave my commits unsquashed so you can read incremental diffs, but once we get something we're happy with I'll rebase/cleanup the git history prior to merging.)

I've updated with the comments.  As alluded to in https://github.com/NixOS/nixpkgs/pull/87440#discussion_r422668974, I think there's some work to be done on the lower half of the manual as well, particularly around how to handle overlays and the examples involved.  Right now it feels a little convoluted and I'd like to highlight the most standard use cases first; but ideally I'd like to do that in a follow-up PR focused on that topic and keep this one focused on the new user onboarding tutorial.  In the meantime I've fixed the blatant error on `numpy`'s version, but it's not a particularly useful example at the moment.

Ma27 commented on doc/languages-frameworks/python.section.md:
````diff
@@ -36,139 +36,208 @@ The Nix and NixOS manuals explain how packages are generally installed. In the
-```nix
-with import <nixpkgs> {};
+Applications on Nix are typically installed into your user profile imperatively
+using `nix-env -i`, and on NixOS declaratively by adding the package name to
````
(nitpick) I'd use `nix-env -iA` here, especially beginners shouldn't even get used to `-i`-only IMHO.

FRidh replied:
If a file does not contain an attribute set, but only a top-level derivation you need to use `nix-env -i -f ...`.

Ma27 commented on doc/languages-frameworks/python.section.md:
```diff
@@ -497,8 +615,8 @@ buildPythonPackage rec {
-It takes an argument `buildPythonPackage`.
-We now call this function using `callPackage` in the definition of our environment
+It takes an argument `buildPythonPackage`. We now call this function using
+`callPackage` in the definition of our environment
```
shouldn't we suggest `python3Packages.callPackage` here?

jonringer replied:
not necessarily, I think the example is clear enough with out having to worry about which interpreter package set your using `callPackage` from.

Ma27 commented on doc/languages-frameworks/python.section.md:
```diff
@@ -976,7 +1093,7 @@ example of such a situation is when `py.test` is used.
-  runs `python setup.py test`. Most python modules follows the standard test
+  runs `python setup.py test`. Most Python modules follows the standard test
   protocol where the pytest runner can be used instead. `py.test` supports a
   `-k` parameter to ignore test methods or classes:
```
Additionally, I'd note that in some cases it's fine to set `doCheck` to `false` (e.g. if there are no tests or only impure tests), but in that case a comment with a rationale should be left above the `doCheck`-assignment.

FRidh replied:
I think that's mentioned in the contributing part.

bhipple commented:
I think this is a shippable improvement over the status quo, so I cleaned up the git history and propose merging. Not sure if I'll have time for a few days, but in a follow up I think we can make further improvements around the bottom section regarding overlays.

jonringer commented:
I took a slightly different approach in my python video: https://www.youtube.com/watch?v=jXd-hkP4xnU&lc=Ugxi6MkvJXqv0Be2mtx4AaABAg and just used the raw python packages.

But your changes LGTM :)
